arangod/Cluster/RebootTracker.cpp

```diff
@@ -66,296 +84,135 @@ RebootTracker::RebootTracker(RebootTracker::SchedulerPointer scheduler)
 #endif
 }
 
-void RebootTracker::updateServerState(
-    containers::FlatHashMap<ServerID, RebootId> const& state) {
-  MUTEX_LOCKER(guard, _mutex);
-
+void RebootTracker::updateServerState(State state) {
+  std::lock_guard guard{_mutex};
```

Comment:
@goedderz 
It seems I need to write what is the difference between the old code and the new one, this is how much happens under the mutex.

### Old code:

1. O(#servers) for each for all elements in old cluster state (std::unordered_map)
2. O(1) find server in new cluster state (absl::flat_hash_map)
3. O(1) find server in _callbacks (std::unordered_map)

then or
1. `scheduleAllCallbacksFor(oldRebootId, max)`
2. then strange O(1) code 
```C++
    auto it = _callbacks.find(serverId);
      if (it != _callbacks.end()) {
        TRI_ASSERT(it->second.empty());
        _callbacks.erase(it);
      }
```
instead of simple `_callbacks.erase(serverId)`

3. O(1) erase from old cluster state (std::unordered_map)

or
1. if oldRebootId < currRebootId call `scheduleAllCallbacksFor(oldRebootId, currRebootId)`

then

1. O(#servers) for each for all elements in new cluster state (std::unordered_map)
2. try to emplace new elements to old map (1-2 allocation per element + O(1) insert + allocations for nodes)


About `scheduleAllCallbacksFor`: O(log(n)) + O(needed reboot id)
1. O(1) find in _callbacks
2. O(log(#reboots id callbacks)) lower_bound
3. O(needed #reboots id callbacks) distance
4. allocate vector
3. O(needed #reboots id callbacks) push to vector
4. O((needed #reboots id callbacks) + log(#reboots id callbacks)) erase or O(1) (if all)

### New code

 1. O(#callbacks servers) iterate for all _callbacks (absl::flat_hash_map)
 2. O(1) find server in new cluster state (absl::flat_hash_map)

if didn't find (common case)
1. O(1) just move map from _callbacks to scheduler (zero allocation)
2. O(1) erase map from _callbacks

if found:
1. O(1) find server in old cluster state (absl::flat_hash_map)
3. if newRebootId and oldRebootId same just next iteration
4. if newRebootId newer than all rebootIds for callbacks (common case) see "didn't find"
5. if we are here just iterate for map and extract all needed node to vector O(#reboots id callbacks) + O(log(#reboots id callbacks)) (we can make single allocation, but I think difficult to compute the difference and simpler code is better here (not common case))

then

1. O(1) move new state to old

Reply:
Also I think new code is simpler to understand, because it makes only necessary things and its smaller (361 vs 218 + 142 vs 106)

Reply:
And I definitely not agree about `unsafe` point. This code have same exception safety as old